src/controllers/hid/hidiothread.cpp

```diff
@@ -0,0 +1,298 @@
+#include "controllers/hid/hidiothread.h"
+
+#include <hidapi.h>
+
+#include <QTimer>
+
+#include "controllers/defs_controllers.h"
+#include "controllers/hid/legacyhidcontrollermappingfilehandler.h"
+#include "moc_hidiothread.cpp"
+#include "util/string.h"
+#include "util/time.h"
+#include "util/trace.h"
+
+namespace {
+constexpr int kReportIdSize = 1;
+constexpr int kMaxHidErrorMessageSize = 512;
+QString loggingCategoryPrefix(const QString& deviceName) {
+    return QStringLiteral("controller.") +
+            RuntimeLoggingCategory::removeInvalidCharsFromCategory(deviceName.toLower());
+}
+} // namespace
+
+HidIoThread::HidIoThread(
+        hid_device* pHidDevice, const mixxx::hid::DeviceInfo& deviceInfo)
+        : QThread(),
+          m_deviceInfo(deviceInfo),
+          // Defining RuntimeLoggingCategories locally in this thread improves runtime performance significiantly
+          m_logBase(loggingCategoryPrefix(deviceInfo.formatName())),
+          m_logInput(loggingCategoryPrefix(deviceInfo.formatName()) +
+                  QStringLiteral(".input")),
+          m_logOutput(loggingCategoryPrefix(deviceInfo.formatName()) +
+                  QStringLiteral(".output")),
+          m_pHidDevice(pHidDevice),
+          m_lastPollSize(0),
+          m_pollingBufferIndex(0) {
+    // Initializing isn't strictly necessary but is good practice.
+    for (int i = 0; i < kNumBuffers; i++) {
+        memset(m_pPollData[i], 0, kBufferSize);
+    }
+    m_outputReportIterator = m_outputReports.begin();
+    m_state.storeRelease(static_cast<int>(HidIoThreadState::Initialized));
+}
+
+HidIoThread::~HidIoThread() {
+    hid_close(m_pHidDevice);
+}
+
+void HidIoThread::run() {
+    while (!testAndSetThreadState(HidIoThreadState::StopRequested, HidIoThreadState::Stopped)) {
+        // Ensure that all InputReports are read from the ring buffer, before the next OutputReport blocks the IO again
+        // Polling available Input-Reports is a cheap software only operation, which takes insignificiant time
+        pollBufferedInputReports();
+
+        // Send one OutputReport, if at least one is latched
+        // Sending an OutputReport is time consuming, because HIDAPI waits for the backend/kernel for confirmation of success
+        // Depending on the OS this takes several several milli seconds
+        // This operation doesn't take many CPU cycles, most time HIDAPI is in idle state
+        if (!sendNextOutputReport()) {
+            if (testAndSetThreadState(HidIoThreadState::StopWhenAllReportsSent,
+                        HidIoThreadState::Stopped)) {
+                return;
+            }
+            usleep(250); // Sleep run loop, if no OutputReport was send
```

Comment:
Could we extract all those magic numbers as constants with descriptive names in an anonymous namespace on the top of the file? We might need to tune those constants later.